### PR TITLE
Tutorial for custom C++ classes in TorchScript

### PR DESCRIPTION
**Pull request by jamesr66a**

### CODE REVIEW

File: advanced_source/torch_script_custom_classes.rst [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+  # Returning and passing instances of custom classes works as you'd expect
+
+  s2 = s.clone()
+  s.merge(s2)
+  for expected in ["bar", "foo", "bar", "foo"]:
+      assert s.pop() == expected
+
+  # We can also use the class in TorchScript
+  # For now, we need to assign the class's type to the local in order to
+  # annotate the type on the TorchScript function
```
Comment (driazati) [inline advanced_source/torch_script_custom_classes.rst]: https://github.com/pytorch/pytorch/pull/29623 might fix this

File: advanced_source/torch_script_custom_classes.rst [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+  Stack = torch.classes.Stack
+
+  # This demonstrates:
```
Comment (driazati) [inline advanced_source/torch_script_custom_classes.rst]: Can you distribute these comments to where these things are actually happening?

File: advanced_source/torch_script_custom_classes.rst [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+  #   - instantiating a class in TorchScript
+  #   - calling a custom class's methods in torchscript
+  #   - returning a custom class instance from TorchScript
+  @torch.jit.script
+  def do_stacks(s : Stack):
+      s2 = torch.classes.Stack(["hi", "mom"])
+      s2.merge(s)
+      return s2.clone(), s2.top()
+
+  stack, top = do_stacks(torch.classes.Stack(["wow"]))
```
Comment (driazati) [inline advanced_source/torch_script_custom_classes.rst]: did you know "wow" is just "mom" upside down

File: advanced_source/torch_script_custom_classes.rst [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+please have a look at the `Loading a TorchScript Model in C++ Tutorial <https://pytorch.org/tutorials/advanced/cpp_export.html>`_.
+
+Similarly to before, let's create a file structure containing the following::
+
+  cpp_inference_example/
+    infer.cpp
+    CMakeLists.txt
+    foo.pt
+    build/
+    custom_class_project/
```
Comment (driazati) [inline advanced_source/torch_script_custom_classes.rst]: Can you fully specify what's in `custom_class_project/` just so everything is explicit?

File: advanced_source/torch_script_custom_classes.rst [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+
+Notice we've copied over the serialized ``foo.pt`` file, as well as the source
+tree from the ``custom_class_project`` above. We will be adding the
+``custom_class_project`` as a dependency to this C++ project so that we can
+build the custom class into the binary.
+
+Let's populate ``infer.cpp`` with the following:
+
+.. code-block:: cpp
```
Comment (driazati) [inline advanced_source/torch_script_custom_classes.rst]: the formatting is broken (codeblock needs an extra leading space)

File: advanced_source/torch_script_custom_classes.rst [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+  find_package(Torch REQUIRED)
+
+  add_subdirectory(custom_class_project)
+
+  # Define our library target
+  add_executable(infer infer.cpp)
+  set(CMAKE_CXX_STANDARD 14)
+  # Link against LibTorch
+  target_link_libraries(infer "${TORCH_LIBRARIES}")
+  target_link_libraries(infer -Wl,--no-as-needed custom_class)
```
Comment (driazati) [inline advanced_source/torch_script_custom_classes.rst]: Add a comment about how this links in the code for `Stack`

File: advanced_source/torch_script_custom_classes.rst [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+    }
+
+    std::vector<c10::IValue> inputs = {"foobarbaz"};
+    auto output = module.forward(inputs).toString();
+    std::cout << output->string() << std::endl;
+  }
+
+And similarly let's define our CMakeLists.txt file:
+
+.. code-block: cmake
```
Comment (driazati) [inline advanced_source/torch_script_custom_classes.rst]: also missing a space

File: advanced_source/torch_script_custom_classes.rst [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+  In this instance, we've passed the specialization of the Stack class ``Stack<std::string>``.
+  In general, you cannot register a non-specialized template class. For non-templated classes,
+  you can just pass the class name directly as the template parameter.
+- The single parameter to ``torch::jit::class_()`` is a string indicating the name of the class.
+  This is the name the class will appear as in both Python and TorchScript. For example, our
+  Stack class would appear as ``torch.classes.Stack``.
+- For each method of the class we'd like to expose to Python and TorchScript, we use the
+  ``.def()`` method on ``torch::jit::class_``. We can chain these together to register
+  multiple methods as well. Let's examine the different callsites of ``def()`` in our example:
+  - ``torch::jit::init<std::vector<std::string>>()`` registers the contructor of our Stack
```
[the remark this answers is not in the document]
Comment (jamesr66a) [reply advanced_source/torch_script_custom_classes.rst]: No. Can you link some docs describing that?

File: advanced_source/torch_script_custom_classes.rst [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+
+Implementing and Binding the Class in C++
+-----------------------------------------
+
+For this tutorial, we are going to define a simple C++ class that maintains persistent
+state in a member variable.
+
+.. code-block:: cpp
+
+  #include <torch/custom_class.h>
```
[the remark this answers is not in the document]
Comment (jamesr66a) [reply advanced_source/torch_script_custom_classes.rst]: Maybe?

**Comment by jamesr66a** [comment]
@dzhulgakov yeah my plan is to work on serialization today

File: advanced_source/torch_script_custom_classes.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (driazati) [reply advanced_source/torch_script_custom_classes.rst]: Was just wondering, it looks [like this](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/init.cpp#L453)

**Pull request merged**